src/stream.rs

```diff
@@ -0,0 +1,293 @@
+use crate::rpc::dcs::unit_update::{UnitGone, Update};
+use crate::rpc::dcs::units_server::Units;
+use crate::rpc::dcs::{
+    GetGroupsRequest, GetUnitsRequest, Position, StreamUnitsRequest, Unit, UnitName,
+};
+use crate::rpc::RPC;
+use futures_util::stream::StreamExt;
+use tokio::sync::mpsc::error::SendError;
+use tokio::sync::mpsc::Sender;
+use tokio::time::MissedTickBehavior;
+use tonic::{Code, Request, Status};
+
+/// Stream unit updates.
+pub async fn stream_units(
+    opts: StreamUnitsRequest,
+    rpc: RPC,
+    tx: Sender<Result<Update, Status>>,
+) -> Result<(), Error> {
+    // initialize the state for the current units stream instance
+    let poll_rate = opts.poll_rate.unwrap_or(5);
+    let max_backoff = Duration::from_secs(opts.max_backoff.unwrap_or(30).max(poll_rate) as u64);
+    let poll_rate = Duration::from_secs(poll_rate as u64);
+    let mut state = State {
+        units: HashMap::new(),
+        ctx: Context {
+            rpc,
+            tx,
+            poll_rate,
+            max_backoff,
+        },
+    };
+
+    // initial full-sync of all current units inside of the mission
+    let groups = state
+        .ctx
+        .rpc
+        .get_groups(Request::new(GetGroupsRequest {
+            category: None,
+            coalition: None,
+        }))
+        .await?
+        .into_inner()
+        .groups;
+    for group in groups {
+        let res = state
+            .ctx
+            .rpc
+            .get_units(Request::new(GetUnitsRequest {
+                group_name: group.name.clone(),
+                active: Some(true),
+            }))
+            .await?
+            .into_inner();
+
+        state.units.extend(
+            res.units
+                .into_iter()
+                .map(|unit| (unit.name.clone(), UnitState::new(unit))),
+        )
+    }
+
+    // send out all initial units
+    for unit_state in state.units.values() {
+        state
+            .ctx
+            .tx
+            .send(Ok(Update::Unit(unit_state.unit.clone())))
+            .await?;
+    }
+
+    // initiate an event stream used to update the state
+    let mut events = state.ctx.rpc.events().await;
+
+    // create an interval used to poll the mission for updates
+    let mut interval = tokio::time::interval(poll_rate);
+    interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+    loop {
+        // wait for either the next event or the next tick, whatever happens first
+        tokio::select! {
+            // listen to events that update the current state
+            Some(crate::rpc::dcs::Event { event: Some(event), .. }) = events.next() => {
+                handle_event(&mut state, event).await?;
+            }
+
+            // poll units for updates
+            _ = interval.tick() => {
+                update_units(&mut state).await?;
+            }
+        }
+    }
+}
+
+/// The state of an active units stream.
+// TODO: re-use one state for all concurrent unit streams?
```

Comment:
I tend towards keeping a separate state per unit stream to not throttle another stream because one of the clients is very slow in consuming its data.

Reply:
> I tend towards keeping a separate state per unit stream to not throttle another stream because one of the clients is very slow in consuming its data.

Are there any advantages to having an mpmc stream? Would it result in less load on the server or something like that? I agree with yout bias of not letting one client screw the others via an accidental "slowloris" attack.

When I wrote tac_scribe, that reads tacview events, I made sure that one thread's job was to receive messages and put them on an internal memory queue for actual processing. This made sure that the receiving thread was fast and would not cause any slowdown.

If there are serious advantages to having mpcp stream then I think we might consider using it and documenting the crap out of the stream system and make sure clients do something similar to tac_scribe but ultimately I am not too fussed either way and defer to you on this one.

Reply:
I gave it some addition thought and I think the question comes down to:

a) Should we expose unit-stream-instance-specific settings (see https://github.com/DCS-gRPC/rust-server/pull/11#discussion_r679436403) or
b) not (= shared state).

When not (b), we can re-use the same state and reduce the amount of calls made to the mission environment in case of concurrent unit streams. I think I could also make it so that one client cannot negatively affect each other (by handling updates very slow), but I'd probably have to buffer updates for each clients and thus terminate them if their respective buffer gets full.

Reply:
I am all for the option that reduces the number of calls made into the mission environment (Which I think is priority #1) and I don't see a problem with re-using the state.

I think the idea of having a buffer and auto-termination is a good idea. One potential issue I can see with auto-termination is that the queue could get instantly full on the initial unit dump as the client connects and therefore the client gets instantly disconnected? Not sure.